govcd/metadata.go

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// GetMetadata() function calls private function getMetadata() with vm.client and vm.VM.HREF
+// which returns a *types.Metadata struct for provided VM input.
+func (vm *VM) GetMetadata() (*types.Metadata, error) {
+	return getMetadata(vm.client, vm.VM.HREF)
+}
+
+// DeleteMetadata() function calls private function deleteMetadata() with vm.client and vm.VM.HREF
+// which deletes metadata depending on key provided as input from VM.
+func (vm *VM) DeleteMetadata(key string) (Task, error) {
+	return deleteMetadata(vm.client, key, vm.VM.HREF)
+}
+
+// AddMetadata() function calls private function addMetadata() with vm.client and vm.VM.HREF
+// which adds metadata key, value pair provided as input to VM.
+func (vm *VM) AddMetadata(key string, value string) (Task, error) {
+	return addMetadata(vm.client, key, value, vm.VM.HREF)
+}
+
+// GetMetadata() function returns meta data for vDC.
+func (vdc *Vdc) GetMetadata() (*types.Metadata, error) {
+	return getMetadata(vdc.client, getAdminVdcURL(vdc.Vdc.HREF))
+}
+
+// DeleteMetadata() function deletes metadata by key provided as input
+func (vdc *Vdc) DeleteMetadata(key string) (Task, error) {
+	return deleteMetadata(vdc.client, key, getAdminVdcURL(vdc.Vdc.HREF))
+}
+
+// AddMetadata() function adds metadata key, value pair provided as input to vDC.
+func (vdc *Vdc) AddMetadata(key string, value string) (Task, error) {
+	return addMetadata(vdc.client, key, value, getAdminVdcURL(vdc.Vdc.HREF))
+}
+
+func getAdminVdcURL(vdcURL string) string {
+	return strings.Split(vdcURL, "/api/vdc/")[0] + "/api/admin/vdc/" + strings.Split(vdcURL, "/api/vdc/")[1]
+}
+
+// GetMetadata() function calls private function getMetadata() with vapp.client and vapp.VApp.HREF
+// which returns a *types.Metadata struct for provided vapp input.
+func (vapp *VApp) GetMetadata() (*types.Metadata, error) {
+	return getMetadata(vapp.client, vapp.VApp.HREF)
+}
+
+func getMetadata(client *Client, requestUri string) (*types.Metadata, error) {
+	metadata := &types.Metadata{}
+
+	_, err := client.ExecuteRequest(requestUri+"/metadata/", http.MethodGet,
+		types.MimeMetaData, "error retrieving metadata: %s", nil, metadata)
+
+	return metadata, err
+}
+
+// DeleteMetadata() function calls private function deleteMetadata() with vapp.client and vapp.VApp.HREF
+// which deletes metadata depending on key provided as input from vApp.
+func (vapp *VApp) DeleteMetadata(key string) (Task, error) {
+	return deleteMetadata(vapp.client, key, vapp.VApp.HREF)
+}
+
+// Deletes metadata (type MetadataStringValue) from the vApp
+// TODO: Support all MetadataTypedValue types with this function
```

Comment:
What are the other types?
What happens if we call this function with unsupported ones?

Reply:
I think @vbauzysvmware inherited string only support. The other types are:
<img width="190" alt="image" src="https://user-images.githubusercontent.com/15804230/60182329-cace2000-982c-11e9-8300-1587be101dc6.png">

But we'd need to alter terraform resources as well to accommodate them. For now we can't fit other types because the variable is string in general.

Reply:
This function existed, I just refactored I moved them to other place. Currently only string supported, so no issue here.